File: PRACTICAS/CAMBIAR_RUBY/RUBY2/Civitas/lib/sorpresa.rb
require_relative 'diario.rb'
require_relative 'casilla.rb'
module Civitas
  class Sorpresa
    
    def initialize(tablero = nil,tipo = nil,valor = 0,texto = "Ninguna",mazo = nil)
      
      @tablero = tablero
      @tipo = tipo
      @valor = valor
      @texto = texto
      @mazo = mazo
      
    end
    
    def self.sorpresa(tablero, tipo)
      self.new(tablero,tipo,-1,nil,nil) 
    end
    
    def self.sorpresa2(tablero, tipo, valor, texto)
      self.new(tablero,tipo,valor,texto,nil) 
    end
    
    def self.sorpresa3(tipo, valor, texto) 
      self.new(nil,tipo,valor,texto,nil) 
    end
    
    def self.sorpresa4(tipo, mazo)
      self.new(nil,tipo,-1,nil,mazo) 
    end
    
    attr_accessor :tipo, :tablero, :valor, :texto
    
    def jugador_correcto(actual, todos)
      return actual < todos.length
    end

    def informe(actual, todos)
      if(jugador_correcto(actual,todos))
        Diario.instance.ocurre_evento("Sorpresa aplicada a: " + todos[actual].nombre)
        puts Diario.instance.leer_evento
      end
    end
    public
    def aplicar_a_jugador(actual, todos)
      case @tipo
      when TipoSorpresa::IRCARCEL
        aplicar_a_jugador_ir_carcel(actual, todos)
      when TipoSorpresa::IRCASILLA
        aplicar_a_jugador_ir_a_casilla(actual, todos)
      when TipoSorpresa::PAGARCOBRAR
        aplicar_a_jugador_pagar_cobrar(actual, todos)
      when TipoSorpresa::PORCASAHOTEL
        aplicar_a_jugador_por_casa_hotel(actual, todos)
      when TipoSorpresa::PORJUGADOR
        aplicar_a_jugador_por_jugador(actual, todos)
      when TipoSorpresa::SALIRCARCEL
        aplicar_a_jugador_salir_carcel(actual, todos)
      end
    end
    
    def aplicar_a_jugador_ir_carcel(actual, todos)
      if (jugador_correcto(actual,todos))
        informe(actual, todos)
        todos[actual].encarcelar(@tablero.num_casilla_carcel)
      end
    end
    
    def aplicar_a_jugador_ir_a_casilla(actual, todos)
      if (jugador_correcto(actual,todos))
        informe(actual, todos)
        casilla_actual = todos[actual].num_casilla_actual
        tirada = @tablero.calcular_tirada(casilla_actual, @valor)
        new_pos = @tablero.nueva_posicion(casilla_actual, tirada)
        todos[actual].mover_a_casilla(new_pos)
        @tablero.get_casilla(new_pos).recibe_jugador(actual, todos)
      end
    end
    
    def aplicar_a_jugador_pagar_cobrar(actual, todos)
      if(jugador_correcto(actual,todos))
        informe(actual, todos)
        todos[actual].modificar_saldo(@valor)
      end
    end
    
    def aplicar_a_jugador_por_casa_hotel(actual, todos)
      if(jugador_correcto(actual, todos))
        informe(actual, todos)
        todos[actual].modificar_saldo(@valor*todos[actual].cantidad_casas_hoteles)
      end
    end
    
    def aplicar_a_jugador_por_jugador(actual, todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        p1 = Sorpresa.new(TipoSorpresa::PAGARCOBRAR, @valor*(-1), "pagar_cobrar")
        
        for i in (0..todos.size - 1)
          if i != actual
            p1.aplicar_a_jugador_pagar_cobrar(i,todos)
          end
          
        end
        
        p2 = Sorpresa.new(TipoSorpresa::PAGARCOBRAR, @valor*(todos.size - 1), "pagar_cobrar")
        p2.aplicar_a_jugador_pagar_cobrar(actual,todos)
      end
    end
    
    def aplicar_a_jugador_salir_carcel(actual, todos)
      if (jugador_correcto(actual, todos))
        informe(actual, todos)
        la_tienen = false
        for i in (0..todos.length-1)
          if (todos.at(i).tiene_salvoconducto)
            la_tienen = true
          end
          if (!la_tienen)
            todos[actual].obtener_salvoconducto(self)
            salir_del_mazo
          end
        end
      end
    end
    
    def salir_del_mazo
      if (@tipo == TipoSorpresa::SALIRCARCEL)
        @mazo.inhabilitar_carta_especial(self)
      end
    end
    
    def usada
      if (@tipo == TipoSorpresa::SALIRCARCEL)
        @mazo.habilitar_carta_especial(self)
      end
    end
    
    def to_s
      return @texto
    end
  end
end

File: PRACTICAS/practica1/Civitas_P1/src/civitas_p1/Sorpresa.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas_p1;


import java.util.ArrayList;

/**
 *
 * @author angel
 */
public class Sorpresa {
    private TipoSorpresa tipo;
    private String texto;
    private int valor;
    private MazoSorpresas mazo;
    private Tablero tablero;
    
    
    Sorpresa(TipoSorpresa _tipo, Tablero tab){
        init();
        tipo = _tipo;
        tablero =tab;
    }
    Sorpresa(TipoSorpresa _tipo, MazoSorpresas maz){
        init();
        tipo = _tipo;
        mazo=maz;
        
    }
    Sorpresa(TipoSorpresa _tipo, Tablero tab, int val, String text){
        init();
        tipo = _tipo;
        tablero =tab;
        valor=val;
        texto=text;
    }
    Sorpresa(TipoSorpresa _tipo, int val, String text){
        init();
        tipo = _tipo;
        valor=val;
        texto=text;
    }
    
    private void init(){
        
        valor=-1;
        mazo=null;
        tablero=null;
        
        
    }
    
    TipoSorpresa getTipo(){
        return tipo;
    }
    
    public boolean jugadorCorrecto(int actual, ArrayList<Jugador>todos){
        
        if(actual>=0 && actual<todos.size()){
            
            return true;
            
        }
        
        return false;
    }
    
    private void informe(int actual, ArrayList<Jugador>todos){
        
        Diario.getInstance().ocurreEvento("Se está aplicando una sorpresa a " + todos.get(actual).getNombre());
        
    }
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        
        if(tipo==TipoSorpresa.IRCARCEL){
            
            aplicarAJugador_irCarcel(actual,todos);
            
        }else if(tipo==TipoSorpresa.IRCASILLA){
            
            aplicarAJugador_irACasilla(actual,todos);
            
        }else if(tipo==TipoSorpresa.PAGARCOBRAR){
            
            aplicarAJugador_pagarCobrar(actual,todos);
            
        }else if(tipo==TipoSorpresa.PORCASAHOTEL){
            
            aplicarAJugador_porCasaHotel(actual,todos);
            
        }else if(tipo==TipoSorpresa.PORJUGADOR){
            
            aplicarAJugador_porJugador(actual,todos);
            
        }else if(tipo==TipoSorpresa.SALIRCARCEL){
            
            aplicarAJugador_salirCarcel(actual,todos);
            
        }
        
    }
    private void aplicarAJugador_irCarcel(int actual, ArrayList<Jugador>todos){
        
        if (jugadorCorrecto(actual,todos)){
            
            informe(actual,todos);
            todos.get(actual).encarcelar(tablero.getCarcel());
            
        }
        
    }
    private void aplicarAJugador_irACasilla(int actual, ArrayList<Jugador>todos){
        
        if (jugadorCorrecto(actual,todos)){
            
            informe(actual,todos);
            int act = todos.get(actual).getNumCasillaActual();
            int tirada = tablero.calcularTirada(act, valor);
            int nueva = tablero.nuevaPosicion(act, tirada);
            todos.get(actual).moverACasilla(nueva);
            tablero.getCasilla(nueva).recibeJugador(actual,todos);
        }
        
    }
    
    private void aplicarAJugador_pagarCobrar(int actual, ArrayList<Jugador> todos){
        
        if(jugadorCorrecto(actual,todos)){
            
            informe(actual,todos);
            todos.get(actual).modificarSaldo(valor);
            
        }
        
    }
    private void aplicarAJugador_porCasaHotel(int actual, ArrayList<Jugador> todos){
        
        if(jugadorCorrecto(actual,todos)){
            
            informe(actual,todos);
            todos.get(actual).modificarSaldo(valor*todos.get(actual).cantidadCasasHoteles());
            
        }
        
    }
    private void aplicarAJugador_porJugador(int actual, ArrayList<Jugador> todos){
        
        if(jugadorCorrecto(actual,todos)){
            
            informe(actual,todos);
            Sorpresa s1 = new Sorpresa(TipoSorpresa.PAGARCOBRAR,valor*(-1),"PagarCobrar");
            for(int i = 0 ; i < todos.size();i++){
                
                if(i!=actual){
                    s1.aplicarAJugador_pagarCobrar(i, todos);
                }
                
            }
            Sorpresa s2=new Sorpresa(TipoSorpresa.PAGARCOBRAR,valor*(todos.size()-1),"PagarCobrar");
            s2.aplicarAJugador_pagarCobrar(actual, todos);
        }
        
    }
    private void aplicarAJugador_salirCarcel(int actual, ArrayList<Jugador> todos){
        
        if(jugadorCorrecto(actual,todos)){
            
            informe(actual,todos);
            boolean tiene=false;
            for(int i = 0; i < todos.size() && !tiene ;i++){
                
                if(todos.get(i).tieneSalvoconducto()){
                    
                    tiene=true;
                    
                }
                
            }
            if(!tiene){
                
                todos.get(actual).obtenerSalvoconducto(this);
                salirDelMazo();
                
            }
            
        }
        
    }
    void salirDelMazo(){
        
        if(tipo==TipoSorpresa.SALIRCARCEL){
            
            mazo.inhabilitarCartaEspecial(this);
            
        }
        
    }
    void usada(){
        
        if(tipo==TipoSorpresa.SALIRCARCEL){
            
            mazo.habilitarCartaEspecial(this);
            
        }
        
    }
    public String toString(){
        
        return texto;
        
    }    
    
}
